sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -686,30 +748,55 @@ public virtual async Task StopProcessingAsync(CancellationToken cancellationToke
         /// specified in the <see cref="MaxConcurrentCalls"/> property.
         /// </summary>
         /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
-        private async Task RunReceiveTaskAsync(
-            CancellationToken cancellationToken)
+        private async Task RunReceiveTaskAsync(CancellationToken cancellationToken)
         {
-            List<Task> tasks = new List<Task>(_maxConcurrentCalls + _receiverManagers.Count);
+            CancellationTokenSource linkedHandlerTcs = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _handlerCts.Token);
+
             try
             {
                 while (!cancellationToken.IsCancellationRequested && !Connection.IsClosed)
                 {
+                    await ReconcileConcurrencyAsync().ConfigureAwait(false);
+
                     foreach (ReceiverManager receiverManager in _receiverManagers)
                     {
+                        // // reset the linkedHandlerTcs if it was already cancelled due to user updating the concurrency
+                        if (linkedHandlerTcs.IsCancellationRequested)
+                        {
+                            linkedHandlerTcs = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _handlerCts.Token);
+                            break;
+                        }
+
                         // Do a quick synchronous check before we resort to async/await with the state-machine overhead.
                         if (!_messageHandlerSemaphore.Wait(0, CancellationToken.None))
                         {
-                            await _messageHandlerSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+                            try
+                            {
+                                await _messageHandlerSemaphore.WaitAsync(linkedHandlerTcs.Token).ConfigureAwait(false);
```

Comment:
You would break the loop twice if linkedHandlerTcs is canceled. Once here and once on the next iteration in `                        if (linkedHandlerTcs.IsCancellationRequested)`